transformations/transformer_fill/README.md

```diff
@@ -0,0 +1,66 @@
+
+Author name: Gautier Dagan
+
+## What type of a transformation is this?
+
+This is a contextual augmentation. This transformation acts as a way to augment the text by changing certain words (according to their POS tag) and replacing `n` of them with words which are most likely under a chose Transformer model.
+
+## What tasks does it intend to benefit?
+
+This augmentation would benefit tasks in low-data settings and can help generalisation. It can also be used to created adversarial examples.
+
+`python evaluate.py -t TransformerFill -task TEXT_CLASSIFICATION`
+`model_name = "aychang/roberta-base-imdb"`
+
+## Previous Work
+
+1. While this code/implementation was written from scratch, I later found that it has some overlap with a previous blog post: https://www.depends-on-the-definition.com/data-augmentation-with-transformers/
+
+2. Augmenting NLP data using Bert's Masked Language capabalities is not new. Multiple works have suggested doing so and showed that it can be used to augment low-resource settings (Wu et. al., 2018; Kumar et. al., 2020) and even create adversarial examples (Garg and Ramakrishnan, 2020).
+
+```bibtex
+@inproceedings{wu2019conditional,
+  title={Conditional bert contextual augmentation},
+  author={Wu, Xing and Lv, Shangwen and Zang, Liangjun and Han, Jizhong and Hu, Songlin},
+  booktitle={International Conference on Computational Science},
+  pages={84--95},
+  year={2019},
+  organization={Springer}
+}
+```
+
+````bibtex
+@article{kumar2020data,
+  title={Data augmentation using pre-trained transformer models},
+  author={Kumar, Varun and Choudhary, Ashutosh and Cho, Eunah},
+  journal={arXiv preprint arXiv:2003.02245},
+  year={2020}
+}
+
+```bibtex
+@article{garg2020bae,
+  title={Bae: Bert-based adversarial examples for text classification},
+  author={Garg, Siddhant and Ramakrishnan, Goutham},
+  journal={arXiv preprint arXiv:2004.01970},
+  year={2020}
+}
+````
+
+## What are the limitations of this transformation?
+
+There are many limitations to such approach and any use of this augmentation will need to be carefully monitored and guided for the task at hand.
+
+1. It is possible that the new augmented sentence has a different meaning from the original, since it is highly dependent on what words the model is asked to replace.
+
+   For instance the sentence `"This is pretty cool"`. If you want choose to replace adjectives, you might end up asking the model to fill in the blank for the sentence: `"This is pretty <mask>"`. In which case it might decide that the most likely word would make: `"This is pretty bad"`.
+
+   A way to alleviate this is to condition the augmentation or prime the model towards certain words. For instance, by prepending `"Positive review:"` to the aformentioned sentence, we are more likely to obtain predictions that will also align with the prompt or primer.
```

Comment:
Hi @gautierdag, thanks for you changes. Have you been able to test this?

Reply:
Similar question as @kaustubhdhole --- I'm not that familiar with distillbert to know if this kind of prefix effect require specific finetuning, but if it's a quantifiable effect (or if there's some paper doing it), then it'd be really cool! other than this, I'm happy to approve the PR.

Reply:
Thank you both for the feedback 😄! I'm currently travelling but will come back to it and add additional tests in the next two weeks.

Reply:
Hey guys, sorry it took me some time to come back to this. 

I've just added a few extra test cases that show priming the model does impact the replaced words.

In addition, I ran the evaluation code provided to measure the impact of the accuracy drop on the IMDB text classification model (`aychang/roberta-base-imdb`).

I use two `TransformerFill` objects to transform the positive and negative examples differently (with different contexts).

```python
from transformations.transformer_fill import TransformerFill

positive_tf = TransformerFill(context_text="POSITIVE REVIEW:", pos_tokens=["ADJ"])
negative_tf = TransformerFill(context_text="NEGATIVE REVIEW:", pos_tokens=["ADJ"])
```

And obtained the following:

```
Non-transformed dataset:
The accuracy on this subset which has 5000 examples= 95.74
Transformed set accuracy:
The accuracy on this subset which has 4999 examples= 95.79915983196639
```

The accuracy slightly increases on the augmented set because adjectives are being switched out for positive/negative counterparts which might align with the target label. 

In doing so I also noticed that long sentences (outside the model max length) would need to be specially truncated due to the variability of the context and the replacement of words -> `<mask>`. So that is now fixed, and it now works out of the box for long texts as well.

@tongshuangwu As to your question about whether the prefix requires fine tuning: In the Wu et. al. 2019 paper on Conditional Bert Contextual Augmentation, they test with and without fine tuning and while fine tuning a LM for the augmentation was best, the naive BERT augmentation still outperformed more traditional augmentation methods such as synonyms.

Let me know if this is enough for you both, I hope to find the time to run additional experiments using the augmented set to train models in low data settings - because this is where this technique should really shine.

Reply:
Thanks for the clarification! Happy to approve on my end.